### PR TITLE
feat(libs): add blockscout-chains crate

### PR DESCRIPTION
**Pull request by lok52**

### CODE REVIEW

File: libs/blockscout-chains/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    let retry_policy = ExponentialBackoff::builder().build_with_max_retries(max_retries);
+    let client = ClientBuilder::new(reqwest::Client::new())
+        .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+        .build();
+    let res = client.get(CHAINS_URL).send().await?;
+    let chains: BlockscoutChains = res.json().await?;
+    Ok(chains)
+}
+
+pub type BlockscoutChains = HashMap<i64, BlockscoutChainData>;
```
Comment (rimrakhimov) [inline libs/blockscout-chains/src/lib.rs:19]: Maybe represent `chain_id` as a `u64` instead?

File: libs/blockscout-chains/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct BlockscoutChainData {
+    pub name: String,
+    pub description: String,
+    pub ecosystem: Ecosystem,
+    pub is_testnet: Option<bool>,
```
Comment (rimrakhimov) [inline libs/blockscout-chains/src/lib.rs:27]: Nit: just interested, what is the case when `is_testnet` is missing?

File: libs/blockscout-chains/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+        .build();
+    let res = client.get(CHAINS_URL).send().await?;
+    let chains: BlockscoutChains = res.json().await?;
+    Ok(chains)
+}
+
+pub type BlockscoutChains = HashMap<i64, BlockscoutChainData>;
+
+#[derive(Deserialize, Debug)]
```
Comment (rimrakhimov) [inline libs/blockscout-chains/src/lib.rs:21]: May we also add `Clone, PartialEq, Eq`, so it be more generic?

File: libs/blockscout-chains/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#[serde(rename_all = "camelCase")]
+pub struct BlockscoutChainData {
+    pub name: String,
+    pub description: String,
+    pub ecosystem: Ecosystem,
+    pub is_testnet: Option<bool>,
+    pub layer: Option<u8>,
```
Comment (rimrakhimov) [inline libs/blockscout-chains/src/lib.rs:28]: Nit: also, when it may be missing?

File: libs/blockscout-chains/src/lib.rs (hunk 2 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs:27]: Just copied that from the original interface:
https://github.com/blockscout/chainscout/blob/7657da41a7dda8c180cfb9ee2fc858c3d27c2dfc/types/index.ts#L12

File: libs/blockscout-chains/src/lib.rs (hunk 4 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs:28]: Even though the original interface specifies this field to be always present, I found some chains where it is `null`.
You can check the response of https://chains.blockscout.com/api/chains and search for `"layer": null`

File: libs/blockscout-chains/src/lib.rs (hunk 1 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs]: Thanks, updated

File: libs/blockscout-chains/src/lib.rs (hunk 4 above)
Comment (rimrakhimov) [reply libs/blockscout-chains/src/lib.rs:28]: Get it, thanks. Let's maybe ask them to fix that inside their interface as well?

File: libs/blockscout-chains/src/lib.rs (hunk 2 above)
Comment (rimrakhimov) [reply libs/blockscout-chains/src/lib.rs:27]: Oh, I see. May we also ask Max why it was the case?

**Review by rimrakhimov — APPROVED** [review]

File: libs/blockscout-chains/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
+use reqwest_middleware::ClientBuilder;
+use reqwest_retry::{policies::ExponentialBackoff, RetryTransientMiddleware};
+use serde::Deserialize;
+use std::collections::HashMap;
+
+const CHAINS_URL: &str = "https://chains.blockscout.com/api/chains";
+
+pub async fn get_blockscout_chains() -> anyhow::Result<BlockscoutChains> {
```
Comment (sevenzing) [inline libs/blockscout-chains/src/lib.rs:8]: why anyhow?

Comment (sevenzing) [reply libs/blockscout-chains/src/lib.rs:8]: maybe use tokio error?

File: libs/blockscout-chains/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    let max_retries = 3;
```
Comment (sevenzing) [inline libs/blockscout-chains/src/lib.rs:9]: nit: write builder to have access to max_retries and chains_url, like

```
let chains = BlockscoutChainsClient::default().fetch_all().await.unwrap()
```
or 
```
let chains = BlockscoutChainsClient::builder().with_max_retries(3).with_host(HOST).build().unwrap().fetch_all().await.unwrpa()
```

File: libs/blockscout-chains/src/lib.rs (hunk 2 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs:76]: Decided to leave the `Option` type as was discussed on the call

File: libs/blockscout-chains/src/lib.rs (hunk 4 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs:77]: Decided to leave the Option type as was discussed on the call

File: libs/blockscout-chains/src/lib.rs (hunk 5 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs]: Used `reqwest_middleware::Error` instead

File: libs/blockscout-chains/src/lib.rs (hunk 6 above)
Comment (lok52) [reply libs/blockscout-chains/src/lib.rs]: Done

**Pull request merged**